compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -683,6 +683,11 @@ private boolean isAssignable(BType source, BType target, Set<TypePair> unresolve
             return false;
         }
 
+        if (isNeverTypeOrStructureTypeWithARequiredNeverMember(source) &&
+                isNeverTypeOrStructureTypeWithARequiredNeverMember(target)) {
+            return true;
+        }
```

Comment:
The following fails at the L648 mutability check.

```ballerina
function func() returns record {| never x; |} {
    panic error("error!");
}

function () returns never x = func;
```

We can maybe repeat this check within that block or make this one of the first checks.

One other option is to analyze the structure type and if it is equivalent to never, set the read-only flag. This would have to be done once all the fields are defined but before semantic analysis I guess.